iceoryx_posh/include/iceoryx_posh/popo/result.hpp

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2021 Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "iceoryx_utils/cxx/expected.hpp"
+
+namespace iox
+{
+namespace popo
+{
+template <typename T, typename Status>
+class Result : public iox::cxx::expected<T, Status>
```

Comment:
I strongly recommend not to inherit from `cxx::expected`. This class does not have a virtual destructor and will never get one.
Additionally, I do not see the benefit of this class but if you intend to have some efficient construction implemented then please be aware that they are already available:
`cxx::expected::create_value` and `cxx::expected::create_error` which are static factory methods for exactly that use case.

If you pursue the approach with the constructor you have to handle the use case where type `T` and type `Status` are equal.

I recommend that we delete this class and add 2 additional constructors to the expected. To differentiate the error and the value type when they are equal we can use the empty struct approach, like this:
```cpp
struct CreateFromValue_t {};
static constexpr CreateFromValue_t CreateFromValue{};
struct CreateFromError_t {};
static constexpr CreateFromError_t CreateFromError{};

template<typename ValueType, typename ErrorType> 
class expected {
  public:
    template<typename... Targs>
    expected(CreateFromValue_t, Targs&&... valueCtorArgs) noexcept;
    template<typename... Targs>
    expected(CreateFromError_t, Targs&&... errorCtorArgs) noexcept;
};
```
I think this is then as efficient as possible and we could also then get rid of the methods `cxx::expected::create_value/error`

Reply:
This was just an experiment and will be removed anyway. Virtual Dtor is only needed when using virtual override, which I never intended to.I was looking for a useful way to augment it for a specific use case without much rewriteing, and for this inheritance like this is fine.